### PR TITLE
discussion-notifications

### PR DESCRIPTION
**Pull request by patte**
frontend todos:
- [x] notifications in setDiscussionPreferences [graphiql](http://localhost:5000/graphiql?query=mutation%20%7B%0A%20%20setDiscussionPreferences(%0A%20%20%20%20id%3A%20%225a8082ea-4b41-427c-8074-32a8d1dea827%22%0A%20%20%20%20discussionPreferences%3A%20%7B%0A%20%20%20%20%20%20anonymity%3A%20false%0A%20%20%20%20%20%20notifications%3A%20MY_CHILDREN%0A%20%20%20%20%7D%0A%20%20)%20%7B%0A%20%20%20%20id%0A%20%20%20%20title%0A%20%20%09userPreference%20%7B%0A%20%20%20%20%20%20notifications%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A)
- [x] updateNotificationSettings for user [query](http://localhost:5000/graphiql?query=%7B%0A%20%20me%20%7B%0A%20%20%20%20defaultDiscussionNotificationOption%0A%20%20%20%20discussionNotificationChannels%0A%20%20%7D%0A%7D%0A) [mutation](http://localhost:5000/graphiql?query=mutation%20%7B%0A%20%20updateNotificationSettings(%0A%20%20%20%20defaultDiscussionNotificationOption%3A%20ALL%0A%20%20%20%20discussionNotificationChannels%3A%20%5BEMAIL%20WEB%5D%0A%20%20)%20%7B%0A%20%20%20%20email%0A%20%20%20%20defaultDiscussionNotificationOption%0A%20%20%20%20discussionNotificationChannels%0A%20%20%7D%0A%7D%0A)
- [x] for [Notifications](https://developer.mozilla.org/en-US/docs/Web/API/notification) we have to ask for permission first. Ask for permission on first comment of user. If declined, remember choice in localstorage to not ask again. If permission is given, for signedIn users [subscribe](http://localhost:5000/graphiql?query=subscription%20%7B%0A%20%20webNotification%20%7B%0A%20%20%20%20title%0A%20%20%20%20body%0A%20%20%20%20icon%0A%20%20%7D%0A%7D%0A) to webNotifications globally and show Notification on update.
- [x] honour `https://republik.ch/url/to/diskussion?mute=1` -> mutate setDiscussionPreferences notifications: NONE

everything up for discussion if frontend has different/further requirements

screen shot 2018-03-07 at 14 10 55
screen shot 2018-03-07 at 14 11 06
screen shot 2018-03-07 at 14 10 40
screen shot 2018-03-07 at 14 15 58
screen shot 2018-03-07 at 14 16 08
screen shot 2018-03-08 at 12 02 24
screen shot 2018-03-08 at 13 05 53

### CODE REVIEW

File: servers/republik/lib/Notifications.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      .map(u => {
+        const user = transformUser(u)
+        return sendMailTemplate({
+          to: u.email,
+          fromEmail: DEFAULT_MAIL_FROM_ADDRESS,
+          fromName: DEFAULT_MAIL_FROM_NAME,
+          subject: t('api/comment/notification/new/email/subject', { discussionName: discussion.title }),
+          templateName: 'cf_comment_notification_new',
+          globalMergeVars: [
+            { name: 'NAME',
```
Comment (tpreusse) [inline servers/republik/lib/Notifications.js]: Add explanation variable `NOTIFICATION_REASON`:
- `Sie erhalten diese E-Mail, weil Sie Benachrichtigungen für Antworten auf ihre Beiträge aktiviert haben. Dies ist die Standardeinstellung.`
- `Sie erhalten diese E-Mail, weil Sie Benachrichtigungen für Antworten in der Debatte «X» aktiviert haben.`
- `Sie erhalten diese E-Mail, weil Sie Benachrichtigungen für alle Beiträge in der Debatte «X» aktiviert haben.`
- `Sie erhalten diese E-Mail, weil Sie Benachrichtigungen für alle Beiträge in Debatten an denen Sie teilgenommen haben aktiviert haben.`

File: servers/republik/graphql/resolvers/_mutations/discussion/submitComment.js [hunk 2]
```diff
@@ -85,6 +94,8 @@ module.exports = async (_, args, { pgdb, user, t, pubsub }) => {
       skipUndefined: true
     })
 
+    await notify(comment, discussion, context)
```
Comment (tpreusse) [inline servers/republik/graphql/resolvers/_mutations/discussion/submitComment.js]: sure that this needs to be awaited? How long does it take to notify 100 people?

Comment (tpreusse) [reply servers/republik/graphql/resolvers/_mutations/discussion/submitComment.js]: why not after slack? and also not await slack? IMO if slack is down we don't want to fail a comment submit.

File: servers/republik/lib/Notifications.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    const webUserIds = notifyUsers
+      .filter(u => u.discussionNotificationChannels.indexOf('WEB') > -1)
+      .map(u => u.id)
+
+    if (webUserIds.length > 0) {
+      pubsub.publish('webNotification', { webNotification: {
+        title: t('api/comment/notification/new/web/subject', { discussionName: discussion.title }),
+        body: `${displayAuthor.name}: ${shortBody}`,
+        icon: t('api/comment/notification/new/web/icon'),
```
Comment (tpreusse) [inline servers/republik/lib/Notifications.js]: `displayAuthor.profilePicture || t('api/comment/notification/new/web/icon')` would be nice

File: servers/republik/graphql/resolvers/_mutations/discussion/submitComment.js (hunk 2 above)
Comment (patte) [reply servers/republik/graphql/resolvers/_mutations/discussion/submitComment.js]: sending webNotifications and emails is not awaited in `servers/republik/lib/Notifications.js`. But I added it there and removed it here for more clarity.

**Comment by tpreusse** [comment]
We need the ability to do a `setDiscussionPreferences` only modifying `notifications`. Currently the front-end had to beak in the assumption that the default value for `anonymity` is false. But it actually depends on the discussion. Additionally setting those values when the user didn't do it in the interface could lead to invisible overwrites (interaction / tab race conditions).

This means `anonymity` has to become option in `DiscussionPreferencesInput`. And credential id has to be preserved unless credential is explicitly `null`.

**Pull request merged**